scalalib/src/mill/scalalib/JpackageModule.scala

```diff
@@ -0,0 +1,83 @@
+package mill
+package scalalib
+
+import mill._
+
+/**
+ * Support for building a native package / installer with the `jpackage` tool which comes bundled with JDK 14 and later.
+ * 
+ * The official `jpackage` docs: https://docs.oracle.com/en/java/javase/23/docs/specs/man/jpackage.html
+ */
+trait JpackageModule extends JavaModule {
+
+  /** The application name */
+  def jpackageName: T[String] = T { artifactName() }
+
+  /** The main class to use as the entry point to the native package / installer. */
+  def jpackageMainClass: T[String] = T { finalMainClass() }
+
+  /** 
+   * The type of native package / installer to be created. 
+   * 
+   * Valid values are: 
+   *  "app-image" - any OS 
+   *  "dmg", "pkg" - macOS (native package, installer)
+   *  "exe", "msi" - Windows (native package, installer)
+   *  "rpm", "deb" - Linux
+   * 
+   * If unspecified, defaults to "app-image" which will build a package native to the host platform. 
+   * */
+  def jpackageType: T[String] = T { "app-image" }
+
+  /**
+   * The classpath used for the `jpackage` tool. The first entry needs to be the main jar.
+   * In difference to [[runClasspath]], it contains the built jars of all dependent modules.
+   */
+  def jpackageRunClasspath: T[Seq[PathRef]] = T {
+    val recLocalClasspath = (localClasspath() ++ transitiveLocalClasspath()).map(_.path)
+
+    val runCp = runClasspath().filterNot(pr => recLocalClasspath.contains(pr.path))
+
+    val mainJar = jar()
+    val recJars = transitiveJars()
+
+    mainJar +: (recJars ++ runCp)
+  }
+
+  /** Builds a native package of the main application. */
+  def jpackageAppImage: T[PathRef] = T {
+    // materialize all jars into a "lib" dir
+    val libs = T.dest / "lib"
+    val cp = jpackageRunClasspath().map(_.path)
+    val jars = cp.filter(os.exists).zipWithIndex.map { case (p, idx) =>
+      val dest = libs / s"${idx + 1}-${p.last}"
+      os.copy(p, dest, createFolders = true)
+      dest
+    }
+
+    val appName = jpackageName()
+    val appType = jpackageType()
+    val mainClass = jpackageMainClass()
+    val mainJarName = jars.head.last
+
+    val args: Seq[String] = Seq(
+      "jpackage",
```

Comment:
We should wire up all these CLI tools to the custom JVM versions that users can define https://mill-build.org/mill/fundamentals/configuring-jvm-versions.html. This probably just means using the executable from `this.zincWorker().javaHome()` rather than the one on `PATH`

Reply:
Thanks for the pointer.

I've updated to use `javaHome()`.